_posts/2017-11-14-nix-in-practice-providing-dependencies.md

```diff
@@ -0,0 +1,183 @@
+tutorial is that you have [installed nix][install] on your system. Also we
+won't focus on the details of the nix language and how the syntax works, if you
+are interested in that have a look at [this nix pill][pill basics].
+
+The question we will focus on in this post is: How to provide dependencies (compilers,
+runtimes, etc.) which
+
+- are provided for the development of a single project only
+- have fixed versions on all systems
+- dont change over time
+
+The solution consists of five steps:
+
+- Step 1: Create a `default.nix` file in your projects folder with the following
+  contents.
+
+  ```nix
+  let
+    hostPkgs = import <nixpkgs> {};
+    nixpkgs = (hostPkgs.fetchFromGitHub {
+      owner = "NixOS";
+      repo = "nixpkgs-channels";
+      rev = "!some-revision!";
+      sha256 = "!sha256!";
+    });
+  in
+  with import nixpkgs {};
+  stdenv.mkDerivation {
+    name = "my-shell";
+    buildInputs = [];
+  }
+  ```
+
+- Step 2: Replace `!some-revision!` in your `default.nix` by a commit hash from
+  the [nixpkgs-channels][nixpkgs-channels-nixos-unstable] repository. Using the latest
+  commit from the `nixos-unstable` branch will work fine most of the time.
+
+- Step 3: Run `nix-shell -p nix-prefetch-git --run "nix-prefetch-git --url
+  https://github.com/NixOS/nixpkgs-channels.git --rev !some-revision!"`,
+  where you replace `!some-revision!` by the git revision you chose. It will
+  take some time to download the `nixpkgs-channels` revision from github and
+  then leave you with output which contains:
+
+  ```
+  ...
+  hash is 0vss6g2gsirl2ds3zaxwv9sc6q6x3zc68431z1wz3wpbhpw190p5
+  ...
+  ```
+
+  Replace `!sha256!` in your `default.nix` with this hash.
+
+- Step 4: Look for all the dependencies you need in the
+  [nixos package search][package search] and add them to the `buildInputs`
+  list. You need to add the value from the "attribute name" column.
+  Commas or string delimiters are not necessary.
+
+  An example for a project where you need Java 8 could in the end look
+  like this:
+
+  ```nix
+  let
+    hostPkgs = import <nixpkgs> {};
+    nixpkgs = (hostPkgs.fetchFromGitHub {
+      owner = "NixOS";
+      repo = "nixpkgs-channels";
+      rev = "1e3995d3ea35f9e4f36bcd18235958759deeb64a";
+      sha256 = "1qn7s5369znh9igj5ysrdakkc0hn5wqcmkryv014y65821b87kv4";
+    });
+  in
+  with import nixpkgs {};
+  stdenv.mkDerivation {
+    name = "my-shell";
+    buildInputs = [ openjdk8 ];
+  }
+  ```
+
+- Step 5: Run `nix-shell` in your projects folder. This should drop you into a shell with the
+  dependencies you defined available in `PATH`. Done. You can now commit the
+  `default.nix` and run `nix-shell` on every system that you want to run your
+  project on and it will provide exactly the same dependencies there. If you
+  want to exit the shell, just use `CTRL+D`.
+
+A good use-case for could be you CI system. You can execute commands inside the
```

Comment:
I think "for" doesn't belong here.

Also, `you` -> `your`

Reply:
👍 Good find.